Libplanet/Blocks/Block.cs

```diff
@@ -188,6 +171,36 @@ private Block(RawBlock rb)
         {
         }
 
+        private Block(
+            HashDigest<SHA256> hash,
+            long index,
+            long difficulty,
+            BigInteger totalDifficulty,
+            Nonce nonce,
+            Address? miner,
+            HashDigest<SHA256>? previousHash,
+            DateTimeOffset timestamp,
+            HashDigest<SHA256>? txHash,
+            IEnumerable<Transaction<T>> transactions,
+            HashDigest<SHA256>? preEvaluationHash,
+            HashDigest<SHA256>? stateRootHash
+        )
+        {
+            Index = index;
+            Difficulty = difficulty;
+            TotalDifficulty = totalDifficulty;
+            Nonce = nonce;
+            Miner = miner;
+            PreviousHash = previousHash;
+            Timestamp = timestamp;
+            Hash = hash;
+            PreEvaluationHash = preEvaluationHash ??
+                throw new ArgumentNullException(nameof(preEvaluationHash));
+            StateRootHash = stateRootHash;
```

Comment:
No null check is needed here?

Reply:
`Block<T>.StateRootHash` is `HashDigest<SHA256>?`, so it can be null.  (I don't have ideas about why.)

Reply:
I was wondering why `PreEvaluationHash` throws `ArgumentNullException` when null while `StateRootHash` isn't. And I found that there can be a block with null `StateRootHash` which is intended.

Reply:
@moreal Could you explain what does the situation `Block<T>.StateRootHash = null` mean?

Reply:
@dahlia `Block<T>.StateRootHash = null` means it is using another `IStateStore` implementation not `TrieStateStore`. Or it may be a block to caculate `StateRootHash`. To calculate `StateRootHash`, it should have a block not having `StateRootHash`.

Reply:
@moreal 

1. Even if a `BlockChain<T>` does not use `TrieStateStore`, shouldn't a `Block<T>` in itself anyway calculate/have its `StateRootHash`?  The current way makes `Block<T>`'s behavior to depend on its `BlockChain<T>`, which depends on `Block<T>` again; [it causes circular dependency][1] which is a [code smell][2].
2. Further on the above point, I believe `Block<T>` should not expose its state that “it should calculate its `StateRootHash` but it hasn't been yet at this *moment*.”

[1]: https://en.wikipedia.org/wiki/Acyclic_dependencies_principle
[2]: https://en.wikipedia.org/wiki/Code_smell

Reply:
It would be better if we have a `FIXME` comment here.

Reply:
@dahlia 

I agree with your opinions:

> it causes circular dependency which is a code smell.
> I believe Block<T> should not expose its state that “it should calculate its StateRootHash but it hasn't been yet at this moment.”

Step by step:

1. To calculate and have `StateRootHash` of `Block<T>`, `Block<T>.Mine()` needs to get `IKeyValueStore` as a parameter at least. It can be smoothly resolved. But if the `BlockChain<T>`does not use `TrieStateStore`, since the time, `Block<T>` loses the duty to do that. A block's shape and the `IStateStore` have strong relations since the block has a field to represent a state of a blockchain.
1. I also agree with your opinion but it has something in common with the above descriptions. As you said, to avoid exposing the state, we must force to use some Merkle trie implementations like `TrieStateStore` to every user. I wonder libplanet's direction in the future for it. If it is the direction, I wonder how `IStateStore` can be used. cc @planetarium/libplanet